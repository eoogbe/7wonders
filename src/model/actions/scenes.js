/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const VIEW_WONDER_CHOOSER = 'VIEW_WONDER_CHOOSER';

export const VIEW_WONDER_PANEL = 'VIEW_WONDER_PANEL';

export const VIEW_GAME_BOARD = 'VIEW_GAME_BOARD';

export const START_AGE = 'START_AGE';

export const SCORE_GAME = 'SCORE_GAME';

export const QUIT = 'QUIT';

export const viewWonderChooser = (playerId) => ({
  type: VIEW_WONDER_CHOOSER,
  payload: {
    playerId,
  },
});

export const viewWonderPanel = () => ({
  type: VIEW_WONDER_PANEL,
});

export const viewGameBoard = () => ({
  type: VIEW_GAME_BOARD,
});

export const startAge = () => ({
  type: START_AGE,
});

export const scoreGame = () => ({
  type: SCORE_GAME,
});

export const quit = () => ({
  type: QUIT,
});
