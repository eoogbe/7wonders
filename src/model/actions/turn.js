/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const BUILD_STRUCTURE = 'BUILD_STRUCTURE';

export const BUILD_WONDER = 'BUILD_WONDER';

export const ADD_MONEY = 'ADD_MONEY';

export const SELECT_HAND_CARD = 'SELECT_HAND_CARD';

export const CHOOSE_NEIGHBOR = 'CHOOSE_NEIGHBOR';

export const COPY_GUILD = 'COPY_GUILD';

export const buildStructure = (cardId) => ({
  type: BUILD_STRUCTURE,
  payload: {
    cardId,
  },
});

export const buildFreeStructure = () => ({
  type: BUILD_STRUCTURE,
  payload: {
    isFree: true,
  },
});

export const buildWonder = () => ({
  type: BUILD_WONDER,
});

export const addMoney = () => ({
  type: ADD_MONEY,
});

export const selectHandCard = (cardId, cardIdx) => ({
  type: SELECT_HAND_CARD,
  payload: {
    cardId,
    cardIdx,
  },
});

export const chooseNeighbor = (costType) => ({
  type: CHOOSE_NEIGHBOR,
  payload: {
    costType,
  },
});

export const copyGuild = (cardId) => ({
  type: COPY_GUILD,
  payload: {
    cardId,
  },
});
