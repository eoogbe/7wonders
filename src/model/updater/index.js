/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { concat, decorate, withDefaultState } from 'redux-fp';
import ages from './ages';
import app, { initialAppState } from './app';
import copyAiGuild from './copyAiGuild';
import hands from './hands';
import military from './military';
import neighbors from './neighbors';
import players from './players';
import takeAiTurns from './takeAiTurns';
import trade from './trade';
import wonder from './wonder';

const initialState = {
  ages: [],
  players: {},
  app: initialAppState,
};

export default decorate(
  withDefaultState(initialState),
  concat(
    wonder,
    neighbors,
    takeAiTurns,
    ages,
    hands,
    players,
    military,
    copyAiGuild,
    app,
    trade
  )
);
