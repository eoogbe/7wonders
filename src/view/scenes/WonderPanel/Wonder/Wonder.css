/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.WonderPanelWonder {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: center;
}

.WonderPanelWonder-currentPlayer {
  font-size: 1.25em;
  font-weight: bold;
}

.WonderPanelWonder-name {
  font-size: 1rem;
  display: flex;
  align-items: center;
  width: 10em;
  height: 6em;
  padding: 0.25em 0.5em;
  border-style: solid;
  border-color: #000;
  border-width: 0 1px 0 1px;
  cursor: pointer;
}
