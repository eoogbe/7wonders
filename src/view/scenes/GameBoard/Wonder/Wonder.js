/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { compose } from 'ramda';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from 'model/actions';
import { getCanBuildFree } from 'model/selectors/players';
import { getShields } from 'presentation/military';
import { getMoney } from 'presentation/players';
import { getWonderStages } from 'presentation/wonder';
import './Wonder.css';
import WonderStage from './WonderStage';

const renderStage = (stageActions) => (stage, idx) => (
  <WonderStage
    key={idx}
    {...stageActions}
    {...stage}
    index={idx}
  />
);

const Wonder = () => {
  const money = useSelector(getMoney);
  const shields = useSelector(getShields);
  const canBuildFree = useSelector(getCanBuildFree);
  const stages = useSelector(getWonderStages);

  const dispatch = useDispatch();
  const showWonderCallout = compose(dispatch, actions.showWonderCallout);

  return (
    <div>
      <div className="GameBoardWonder-tokenContainer">
        <div className="GameBoardWonder-token">{money}</div>
        <div className="GameBoardWonder-token">{shields}</div>
        {canBuildFree &&
          <div className="GameBoardWonder-token">Can build free structure</div>}
      </div>
      <ol className="GameBoardWonder-stageList">
        {stages.map(renderStage({ showWonderCallout }))}
      </ol>
    </div>
  );
};

export default Wonder;
