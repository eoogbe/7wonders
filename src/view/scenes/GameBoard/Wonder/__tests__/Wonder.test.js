/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PLAYERS from '__fixtures__/players';
import { showWonderCallout } from 'model/actions';
import Wonder from '../Wonder';

jest.useFakeTimers();

const mockStore = configureStore();

it('renders without crashing', () => {
  const store = mockStore(createState({}));
  const tree =
    renderer.create(
      <Provider store={store}>
        <Wonder />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders when can build a free structure', () => {
  const store = mockStore(createState({
    wonder: 'STATUE_OF_ZEUS_IN_OLYMPIA',
    wonderAbilities: ['BUILD_FREE'],
  }));
  const tree =
    renderer.create(
      <Provider store={store}>
        <Wonder />
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('shows the wonder callout', () => {
  const store = mockStore(createState({}));
  const tree = mount(
    <Provider store={store}>
      <Wonder />
    </Provider>
  );
  tree.find('.GameBoardWonderStage').first().simulate('mouseEnter');
  expect(store.getActions()).toEqual([showWonderCallout(0)]);
});

const createState = (player) => ({
  app: { currentPlayer: 'PLAYER1' },
  players: {
    ...PLAYERS,
    PLAYER1: {
      ...PLAYERS.PLAYER1,
      ...player,
    },
  },
});
