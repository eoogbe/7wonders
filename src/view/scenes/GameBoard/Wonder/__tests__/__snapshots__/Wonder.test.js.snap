// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when can build a free structure 1`] = `
<div>
  <div
    className="GameBoardWonder-tokenContainer"
  >
    <div
      className="GameBoardWonder-token"
    >
      3 coins
    </div>
    <div
      className="GameBoardWonder-token"
    >
      1 shield
    </div>
    <div
      className="GameBoardWonder-token"
    >
      Can build free structure
    </div>
  </div>
  <ol
    className="GameBoardWonder-stageList"
  >
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={Object {}}
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Wood
        </li>
        <li>
          Wood
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          3 points
        </li>
      </ul>
    </li>
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(93, 64, 55, 0.5)",
        }
      }
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Stone
        </li>
        <li>
          Stone
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          Build free
        </li>
      </ul>
      <div
        className="GameBoardWonderStage-payment"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8 0-1.85.63-3.55 1.69-4.9L16.9 18.31C15.55 19.37 13.85 20 12 20zm6.31-3.1L7.1 5.69C8.45 4.63 10.15 4 12 4c4.42 0 8 3.58 8 8 0 1.85-.63 3.55-1.69 4.9z"
          />
        </svg>
      </div>
    </li>
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(93, 64, 55, 0.5)",
        }
      }
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Ore
        </li>
        <li>
          Ore
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          7 points
        </li>
      </ul>
    </li>
  </ol>
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <div
    className="GameBoardWonder-tokenContainer"
  >
    <div
      className="GameBoardWonder-token"
    >
      3 coins
    </div>
    <div
      className="GameBoardWonder-token"
    >
      1 shield
    </div>
  </div>
  <ol
    className="GameBoardWonder-stageList"
  >
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={Object {}}
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Stone
        </li>
        <li>
          Stone
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          3 points
        </li>
      </ul>
    </li>
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(93, 64, 55, 0.5)",
        }
      }
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Wood
        </li>
        <li>
          Wood
        </li>
        <li>
          Wood
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          5 points
        </li>
      </ul>
      <div
        className="GameBoardWonderStage-payment"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8 0-1.85.63-3.55 1.69-4.9L16.9 18.31C15.55 19.37 13.85 20 12 20zm6.31-3.1L7.1 5.69C8.45 4.63 10.15 4 12 4c4.42 0 8 3.58 8 8 0 1.85-.63 3.55-1.69 4.9z"
          />
        </svg>
      </div>
    </li>
    <li
      className="GameBoardWonderStage"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(93, 64, 55, 0.5)",
        }
      }
    >
      <ul
        className="GameBoardWonderStage-cost"
      >
        <li>
          Stone
        </li>
        <li>
          Stone
        </li>
        <li>
          Stone
        </li>
        <li>
          Stone
        </li>
      </ul>
      <ul
        className="GameBoardWonderStage-effectList"
      >
        <li>
          7 points
        </li>
      </ul>
    </li>
  </ol>
</div>
`;
